cpp/src/arrow/compute/kernels/scalar_nested.cc

```diff
@@ -819,6 +832,378 @@ const FunctionDoc map_lookup_doc{
+          }
+        } else {
+          builder->UnsafeAppend(UnboxScalar<BooleanType>::Unbox(*input.scalar));
+        }
+      }
+    }
+    return list_builder.FinishInternal(out->array_data_mutable());
+  }
+
+  // Numeric and temporal types
+  template <typename InputType>
+  std::enable_if_t<has_c_type<InputType>::value || is_temporal_type<InputType>::value,
+                   Status>
+  Visit(const InputType& input_type, KernelContext* ctx, const ExecSpan& batch,
+        ExecResult* out) {
+    using BuilderType = typename TypeTraits<InputType>::BuilderType;
+    using ListBuilderType = typename TypeTraits<OutputType>::BuilderType;
+
+    auto builder = std::make_shared<BuilderType>(input_type.GetSharedPtr(),
+                                                 ctx->exec_context()->memory_pool());
+    ListBuilderType list_builder(ctx->exec_context()->memory_pool(), builder, list_type);
+
+    RETURN_NOT_OK(builder->Reserve(batch.num_values() * batch.length));
+    RETURN_NOT_OK(list_builder.Reserve(batch.length));
+
+    for (int i = 0; i < batch.length; ++i) {
+      RETURN_NOT_OK(list_builder.Append());
+      for (const auto& input : batch.values) {
+        if (input.is_array()) {
+          const auto& arr = input.array;
+          if (arr.IsValid(i)) {
+            builder->UnsafeAppend(arr.GetValues<typename InputType::c_type>(1)[i]);
+          } else {
+            builder->UnsafeAppendNull();
+          }
+        } else {
+          builder->UnsafeAppend(UnboxScalar<InputType>::Unbox(*input.scalar));
+        }
+      }
+    }
+    return list_builder.FinishInternal(out->array_data_mutable());
+  }
+
+  // Varlen binary types
+  template <typename InputType>
+  std::enable_if_t<is_base_binary_type<InputType>::value, Status> Visit(
+      const InputType& input_type, KernelContext* ctx, const ExecSpan& batch,
+      ExecResult* out) {
+    using BuilderType = typename TypeTraits<InputType>::BuilderType;
+    using ListBuilderType = typename TypeTraits<OutputType>::BuilderType;
+    using OffsetType = typename TypeTraits<InputType>::OffsetType::c_type;
+    auto builder = std::make_shared<BuilderType>();
+    ListBuilderType list_builder(ctx->exec_context()->memory_pool(), builder, list_type);
+
+    RETURN_NOT_OK(builder->Reserve(batch.num_values() * batch.length));
+    RETURN_NOT_OK(ReserveBinaryData<InputType>(batch, builder.get()));
+
+    RETURN_NOT_OK(list_builder.Reserve(batch.length));
+
+    for (int i = 0; i < batch.length; ++i) {
+      RETURN_NOT_OK(list_builder.Append());
+      for (const auto& input : batch.values) {
+        if (input.is_array()) {
+          const auto& arr = input.array;
+          if (arr.IsValid(i)) {
+            auto cur_offset = arr.GetValues<OffsetType>(1)[i];
+            auto next_offset = arr.GetValues<OffsetType>(1)[i + 1];
+            std::string_view view(arr.buffers[2].data_as<char>() + cur_offset,
+                                  next_offset - cur_offset);
```

Comment:
It's annoying that we have a `GetView` method on all concrete `Array` classes (numeric, binary) that does the right thing and would allow consolidating several of those Visit methods, but `ArraySpan` does not offer the same facility.

I opened https://github.com/apache/arrow/issues/37153

Reply:
Added concrete array span types for all pritimive types. I'm not sure if unit tests are needed.